lesson6-Reflection/Task1/Container.cs

```diff
@@ -1,28 +1,100 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.Linq;
 using System.Reflection;
+using Task1.DoNotChange;
 
 namespace Task1
 {
     public class Container
     {
+        private readonly Dictionary<Type, Type> _mappedTypes = new Dictionary<Type, Type>();
+
         public void AddAssembly(Assembly assembly)
         {
-            throw new NotImplementedException();
+            var mappedTypes = assembly.GetTypes().Where(t =>
+                Attribute.IsDefined(t, typeof(ExportAttribute)) ||
+                Attribute.IsDefined(t, typeof(ImportConstructorAttribute)) ||
+                t.GetProperties().Any(p => Attribute.IsDefined(p, typeof(ImportAttribute))) ||
+                t.GetFields().Any(f => Attribute.IsDefined(f, typeof(ImportAttribute))));
+
+            foreach (var type in mappedTypes)
+                if (type.GetCustomAttributes(typeof(ExportAttribute)).FirstOrDefault() == null)
+                    AddType(type);
+                else if ((type.GetCustomAttributes(typeof(ExportAttribute)).FirstOrDefault() as ExportAttribute)?.Contract != null)
+                    AddType(type,
+                        (type.GetCustomAttributes(typeof(ExportAttribute)).First() as ExportAttribute)?.Contract);
+                else
+                    AddType(type);
         }
 
         public void AddType(Type type)
         {
-            throw new NotImplementedException();
+            // If Already added, update the injection type
+            if (_mappedTypes.ContainsKey(type))
+                _mappedTypes[type] = type;
+            _mappedTypes.Add(type, type);
         }
 
         public void AddType(Type type, Type baseType)
         {
-            throw new NotImplementedException();
+            //TODO: ask why we should check it
+            // If Already added, update the injection type
+            //if (_mappedTypes.ContainsKey(baseType))
+            //  _mappedTypes[baseType] = type;
+            
+            AddType(type);
+            _mappedTypes.Add(baseType, type);
         }
 
         public T Get<T>()
         {
-            throw new NotImplementedException();
+            if (!_mappedTypes.Any())
+                throw new Exception("No entity has been added yet");
+            return (T)Resolve(typeof(T));
+        }
+
+        private object Resolve(Type type)
+        {
+            Type resolvedType;
+            try
+            {
```

Comment:
Catching and throwing exception is resource-consuming operation, I'd recommend instead to throw exception if the type was not found in dictionary:
```suggestion
            if(!_mappedTypes.ContainsKey(type))
                throw new Exception(...);
```

Reply:
done